### DISCUSSION TITLE
Installation testing error

### DISCUSSION THREAD

**Discussion by Mousa-saleh** [neuropoly/axondeepseg#526] [Help needed] [answered]
Hello all, 
I just installed Axondeepseg and when I tested the installation, I got the following errors:

![image](https://user-images.githubusercontent.com/85881282/121894266-9f011900-cd27-11eb-90d2-f11364591c46.png)

Can any one tell me what is the problem??

Thanks,

**Comment by vasudev-sharma** [comment]
Hey @Mousa-saleh thanks for reporting the issue. Would it be possible for you to paste the **entire logs of the terminal** here so that it would help us to debug the core problem of the issue? 

Additionally, what **installation steps** you followed? This information would be really helpful for us to reproduce the error on our end. Regarding installation steps, I recommend you following the [installation documentation](https://axondeepseg.readthedocs.io/en/latest/documentation.html#installation) of ADS. If the error still persists, let us know.

Also, could you complete the following information?

**Platform:**
 - OS: [e.g. Mac OS "Flaming Flamingo", Fedora 99, Ubuntu 48.04]
 - AxonDeepSeg version number [e.g. 3.2.0, git master, etc.]

**Reply by Mousa-saleh** [reply]
Thanks @vasudev-sharma
I followed the installation steps you mentioned.
![image](https://user-images.githubusercontent.com/85881282/121904115-96addb80-cd31-11eb-8816-5a87004ef149.png)
![image](https://user-images.githubusercontent.com/85881282/121904278-b7763100-cd31-11eb-9212-130a5984e312.png)
I think the problem started when I did the test command (axondeepseg_test)
first, I got the message( using theano backend).
I searched over the web how to convert it to (using tensorflow backend) and found that I have to delete the file (keras_activate) and it worked.
I deactivated the env. and activated it again and got (using tensorflow backend) as shown in the two images above.
then I ran the command (py.test --cov AxonDeepSeg/ --cov-report term-missing) for entire testing and got the 7 errors sent in the first post.

- OS: windows 8.1
- I don't know how to figure the axondeepseg version.

**Reply by Mousa-saleh** [reply]
I have another problem.
I tried to work on the jupyter notebook (00-getting_started)
I got an error that the following directory is not found!!
path_testing = os.path.join('..','AxonDeepSeg','models','default_SEM_model','data_test')

Thanks,

**Reply by mariehbourget** [reply]
> I think the problem started when I did the test command (axondeepseg_test)
> first, I got the message( using theano backend).
> I searched over the web how to convert it to (using tensorflow backend) and found that I have to delete the file (keras_activate) and it worked.

Hi @Mousa-saleh, thank you for reaching out!
Regarding the `Using Theano backend` issue, you will find in our documentation a note in this [section](https://axondeepseg.readthedocs.io/en/latest/documentation.html#quick-test) that explain how to solve this issue by exporting the `KERAS_BACKEND` variable in the `keras_activate.sh` file.

 I suggest you try to re-install ADS in a new conda environment using this fix and let us know if the errors persist.

**Reply by Mousa-saleh** [reply]
Thanks @mariehbourget 
I am not sure where to add the export KERAS_BACKEND="tensorflow" 
I went to the file keras_active and edited it in a notepad 
I removed the word theano and replaced it with tensorflow 
Is this correct??
if not, please help me with the steps
Thanks,

**Reply by mariehbourget** [reply]
You can edit the file in notepad and add the line at the very end without replacing anything.

For example, the content of the `keras_activate.sh` file before the fix:
```
#!/bin/bash

# Figure out the default Keras backend by reading the config file.
KERAS_BACKEND="$(python ${CONDA_PREFIX}/etc/keras/load_config.py)"

# Try to use the default Keras backend.
# Fallback to Theano if it fails (Theano always works).
python -c "import keras" &>/dev/null || {
    test "true";
    export KERAS_BACKEND="theano"
    python -c "import keras" &>/dev/null
}
```

And the content after the fix with the extra `export KERAS_BACKEND="tensorflow"` line at the end:
```
#!/bin/bash

# Figure out the default Keras backend by reading the config file.
KERAS_BACKEND="$(python ${CONDA_PREFIX}/etc/keras/load_config.py)"

# Try to use the default Keras backend.
# Fallback to Theano if it fails (Theano always works).
python -c "import keras" &>/dev/null || {
    test "true";
    export KERAS_BACKEND="theano"
    python -c "import keras" &>/dev/null
}

export KERAS_BACKEND="tensorflow"
```

Once it's done, make sure to deactivate/re-activate your conda environment.

**Reply by Mousa-saleh** [reply]
Is it ok that the content of the file I have is different?
here is what I have 

:: Figure out the default Keras backend by reading the config file.
python %CONDA_PREFIX%\etc\keras\load_config.py > temp.txt
set /p KERAS_BACKEND=<temp.txt
del temp.txt

:: Try to use the default Keras backend.
:: Fallback to Theano if it fails (Theano always works).
python -c "import keras" 1> nul 2>&1
if errorlevel 1 (
    ver > nul
    set "KERAS_BACKEND=tensorflow"
    python -c "import keras" 1> nul 2>&1
)

**Reply by mariehbourget** [reply]
Oh yes, sorry my mistake, the files are slightly different because of OS (Linux vs. Windows)

So in Windows, the file to edit is `keras_activate.bat`, and I think the line to add would be:
`set KERAS_BACKEND="tensorflow"` at the end like in the example.

**Reply by vasudev-sharma** [reply]
> I have another problem.
> I tried to work on the jupyter notebook (00-getting_started)
> I got an error that the following directory is not found!!
> path_testing = os.path.join('..','AxonDeepSeg','models','default_SEM_model','data_test')
> 
> Thanks,

The likely cause of this error is that ADS has not been installed properly.  In the `axondeepseg` directory, could you try running this command in the virtual environment to see if the ADS gets properly installed?

```bash
$ pip install -e .
```

This will install **ADS** and download the model (`default_SEM_model`), which would most likely solve this error of "directory is not found!!".

Let us know how it goes.

**Reply by Mousa-saleh** [reply]
Thanks guys, it worked but I ran into another error when running the following command:
(from AxonDeepSeg.apply_model import axon_segmentation)

the error message:

Using TensorFlow backend.
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
~\.conda\envs\ads_venv\lib\site-packages\tensorflow\python\pywrap_tensorflow.py in <module>
     57 
---> 58   from tensorflow.python.pywrap_tensorflow_internal import *
     59   from tensorflow.python.pywrap_tensorflow_internal import __version__

~\.conda\envs\ads_venv\lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py in <module>
     27             return _mod
---> 28     _pywrap_tensorflow_internal = swig_import_helper()
     29     del swig_import_helper

~\.conda\envs\ads_venv\lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py in swig_import_helper()
     23             try:
---> 24                 _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
     25             finally:

~\.conda\envs\ads_venv\lib\imp.py in load_module(name, file, filename, details)
    241         else:
--> 242             return load_dynamic(name, filename, file)
    243     elif type_ == PKG_DIRECTORY:

~\.conda\envs\ads_venv\lib\imp.py in load_dynamic(name, path, file)
    341             name=name, loader=loader, origin=path)
--> 342         return _load(spec)
    343 

ImportError: DLL load failed: A dynamic link library (DLL) initialization routine failed.

During handling of the above exception, another exception occurred:

ImportError                               Traceback (most recent call last)
<ipython-input-7-a71bb435a888> in <module>
----> 1 from AxonDeepSeg.apply_model import axon_segmentation

c:\users\mousaa\axondeepseg\AxonDeepSeg\apply_model.py in <module>
      6 import AxonDeepSeg.ads_utils as ads
      7 from AxonDeepSeg.ads_utils import convert_path
----> 8 from AxonDeepSeg.network_construction import uconv_net
      9 from AxonDeepSeg.visualization.get_masks import get_masks
     10 from AxonDeepSeg.patch_management_tools import im2patches_overlap, patches2im_overlap

c:\users\mousaa\axondeepseg\AxonDeepSeg\network_construction.py in <module>
----> 1 from keras.layers import (
      2                             Conv2D,
      3                             BatchNormalization,
      4                             Activation,
      5                             Dropout,

~\.conda\envs\ads_venv\lib\site-packages\keras\__init__.py in <module>
      1 from __future__ import absolute_import
      2 
----> 3 from . import utils
      4 from . import activations
      5 from . import applications

~\.conda\envs\ads_venv\lib\site-packages\keras\utils\__init__.py in <module>
      4 from . import data_utils
      5 from . import io_utils
----> 6 from . import conv_utils
      7 from . import losses_utils
      8 from . import metrics_utils

~\.conda\envs\ads_venv\lib\site-packages\keras\utils\conv_utils.py in <module>
      7 from six.moves import range
      8 import numpy as np
----> 9 from .. import backend as K
     10 
     11 

~\.conda\envs\ads_venv\lib\site-packages\keras\backend\__init__.py in <module>
----> 1 from .load_backend import epsilon
      2 from .load_backend import set_epsilon
      3 from .load_backend import floatx
      4 from .load_backend import set_floatx
      5 from .load_backend import cast_to_floatx

~\.conda\envs\ads_venv\lib\site-packages\keras\backend\load_backend.py in <module>
     88 elif _BACKEND == 'tensorflow':
     89     sys.stderr.write('Using TensorFlow backend.\n')
---> 90     from .tensorflow_backend import *
     91 else:
     92     # Try and load external backend.

~\.conda\envs\ads_venv\lib\site-packages\keras\backend\tensorflow_backend.py in <module>
      3 from __future__ import print_function
      4 
----> 5 import tensorflow as tf
      6 from tensorflow.python.eager import context
      7 from tensorflow.python.framework import device as tfdev

~\.conda\envs\ads_venv\lib\site-packages\tensorflow\__init__.py in <module>
     22 
     23 # pylint: disable=g-bad-import-order
---> 24 from tensorflow.python import pywrap_tensorflow  # pylint: disable=unused-import
     25 
     26 from tensorflow._api.v1 import app

~\.conda\envs\ads_venv\lib\site-packages\tensorflow\python\__init__.py in <module>
     47 import numpy as np
     48 
---> 49 from tensorflow.python import pywrap_tensorflow
     50 
     51 # Protocol buffers

~\.conda\envs\ads_venv\lib\site-packages\tensorflow\python\pywrap_tensorflow.py in <module>
     72 for some common reasons and solutions.  Include the entire stack trace
     73 above this error message when asking for help.""" % traceback.format_exc()
---> 74   raise ImportError(msg)
     75 
     76 # pylint: enable=wildcard-import,g-import-not-at-top,unused-import,line-too-long

ImportError: Traceback (most recent call last):
  File "C:\Users\mousaa\.conda\envs\ads_venv\lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "C:\Users\mousaa\.conda\envs\ads_venv\lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "C:\Users\mousaa\.conda\envs\ads_venv\lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
  File "C:\Users\mousaa\.conda\envs\ads_venv\lib\imp.py", line 242, in load_module
    return load_dynamic(name, filename, file)
  File "C:\Users\mousaa\.conda\envs\ads_venv\lib\imp.py", line 342, in load_dynamic
    return _load(spec)
ImportError: DLL load failed: A dynamic link library (DLL) initialization routine failed.


Failed to load the native TensorFlow runtime.

**Reply by Mousa-saleh** [reply]
The problem seems to be that my processor is not AVX supported,
Can I run the code on Google Colab?

**Reply by Mousa-saleh** [reply]
my processor is Intel Core 2 duo

**Reply by mariehbourget** [reply]
Hi @Mousa-saleh, the error you have with `AxonDeepSeg.apply_model import axon_segmentation`  is similar to the first one in your screen shot. I think maybe tensorflow is still not recognized properly.

First, we have to make sure that the installation is working correctly. Let us know if you able to run the `axondeepseg_test` without errors after the modification suggested to` keras_activate`. Thanks!

We will get back to you as soon as possible regarding your other questions about processor and Google Colab.

**Comment by hermancollin** [comment, marked as answer]
Hi @Mousa-saleh. It looks to me like you have some compatibility issues because of the Windows version you are using. Have you considered using a Linux virtual machine, and then install AxonDeepSeg inside it? [Virtualbox](https://www.virtualbox.org/) is a good one if you wish to try this on Windows 8.
https://www.wikihow.com/Install-Ubuntu-on-VirtualBox
It is unlikely that you will encounter any of these problems if you install ADS on a stable linux distribution. 

edit: I should mention this would affect the performance of ADS. Alternatively, [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/about) offers less overhead, but it hasn't been tested by our team on Windows 8.

**Comment by mathieuboudreau** [comment]
Hi @Mousa-saleh, 

As for the possible compatibility with Google Colab, none of us in the AxonDeepSeg team have experience with that. For that to work, you would need to get a conda virtual environment running within a notebook. I tried doing so using [these instructions](https://towardsdatascience.com/conda-google-colab-75f7c867a522), however the final step of installing ADS failed because there was a conflict between the python installed by default in google colab and the python executable in the virtual environment.

See here for my attempt, feel free to download and test it yourself, and try to debug the final issue if you'd like. https://gist.github.com/mathieuboudreau/5fb9a828b1e067f4923cd083961e342b

**Comment by Mousa-saleh** [comment]
Thank you for your responses,

I worked on another laptop and it worked fine 🙂

Thanks,

**Comment by Mousa-saleh** [comment]
Hello guys,

I ran a test image on the getting started notebook.
It seems that the model only detect small neurons.
![1](https://user-images.githubusercontent.com/85881282/122395847-1204cc80-cf80-11eb-9a86-19c4d5fb21b0.PNG)
Can any help me figuring out what is the problem??

Thanks in advance,

**Reply by jcohenadad** [reply]
Hi @Mousa-saleh , [this](https://github.com/neuropoly/axondeepseg/discussions/526#discussioncomment-882960) seems to be a different error than the title of this thread (Installation testing error). For the sake of keeping this discussion Q&A well organized, could you please post this as a new discussion thread with a specific title (eg: "only small axons are detected on the notebook tutorial"). Thank you

**Reply by Mousa-saleh** [reply]
Ok,
Thanks,
